File: src/BrewSpa/Dashboards/BrewSpa.Dashboards.ApplicationServices/Services/DashboardService.cs
using System.Net.Http.Json;
using System.Text;
using BrewSpa.Dashboards.ApplicationServices.Models;
using BrewSpa.Shared.Models;
using Lena.Core;

namespace BrewSpa.Dashboards.ApplicationServices.Services;

internal class DashboardService(HttpClient httpClient) : IDashboardService
{
    public async Task<Result<PagedResult<SalesByCustomerJson>>> GetSalesByCustomerAsync(int pageNumber = 1, int pageSize = 10)
    {
        try
        {
            var requestUri = $"customers?pageNumber={pageNumber}&pageSize={pageSize}";
            var httpResponse = await httpClient.GetAsync(requestUri);
            if (!httpResponse.IsSuccessStatusCode)
            {
                var errorContent = await httpResponse.Content.ReadAsStringAsync();
                var errorMessage = new StringBuilder();
                errorMessage.AppendLine($"[DashboardService] Error Content: {errorContent}");
                errorMessage.AppendLine("[DashboardService] GetSalesByCustomerAsync API call failed");
                return Result<PagedResult<SalesByCustomerJson>>.Error(errorMessage.ToString());
            }

            var response = await httpResponse.Content.ReadFromJsonAsync<PagedResult<SalesByCustomerJson>>();
            return Result<PagedResult<SalesByCustomerJson>>.Success(
                new PagedResult<SalesByCustomerJson>(response!.Results, response.Page, response.PageSize, response.TotalRecords));
        }
        catch (Exception ex)
        {
            var errorMessage = new StringBuilder();
            errorMessage.Append($"[DashboardService] Exception: {ex.Message}");
            errorMessage.Append($"[DashboardService] Stack Trace: {ex.StackTrace}");
            return Result<PagedResult<SalesByCustomerJson>>.Error(errorMessage.ToString());
        }
    }

    public async Task<Result<PagedResult<SalesByProductJson>>> GetSalesByProductAsync(int pageNumber = 1, int pageSize = 10)
    {
        try
        {
            var requestUri = $"products?pageNumber={pageNumber}&pageSize={pageSize}";
            var httpResponse = await httpClient.GetAsync(requestUri);
            if (!httpResponse.IsSuccessStatusCode)
            {
                var errorContent = await httpResponse.Content.ReadAsStringAsync();
                var errorMessage = new StringBuilder();
                errorMessage.AppendLine($"[DashboardService] Error Content: {errorContent}");
                errorMessage.AppendLine("[DashboardService] GetSalesByProductAsync API call failed");
                return Result<PagedResult<SalesByProductJson>>.Error(errorMessage.ToString());
            }

            var response = await httpResponse.Content.ReadFromJsonAsync<PagedResult<SalesByProductJson>>();
            return Result<PagedResult<SalesByProductJson>>.Success(
                new PagedResult<SalesByProductJson>(response!.Results, response.Page, response.PageSize, response.TotalRecords));
        }
        catch (Exception ex)
        {
            var errorMessage = new StringBuilder();
            errorMessage.Append($"[DashboardService] Exception: {ex.Message}");
            errorMessage.Append($"[DashboardService] Stack Trace: {ex.StackTrace}");
            return Result<PagedResult<SalesByProductJson>>.Error(errorMessage.ToString());
        }
    }
}

File: src/BrewApp/src/app/features/dashboards/services/dashboard.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { API_BASE_URL } from '../../chat/services/chat.service';
import { PagedResult } from '../../../shared/models/paged-result.model';
import { SalesByCustomerJson, SalesByProductsJson } from '../models/dashboard.model';

@Injectable({ providedIn: 'root' })
export class DashboardService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = inject(API_BASE_URL);

  getSalesByCustomer(pageNumber = 1, pageSize = 10): Observable<PagedResult<SalesByCustomerJson>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber)
      .set('pageSize', pageSize);
    return this.http.get<PagedResult<SalesByCustomerJson>>(`${this.baseUrl}/v1/dashboards/customers`, { params });
  }

  getSalesByProducts(pageNumber = 1, pageSize = 10): Observable<PagedResult<SalesByProductsJson>> {
    const params = new HttpParams()
      .set('pageNumber', pageNumber)
      .set('pageSize', pageSize);
    return this.http.get<PagedResult<SalesByProductsJson>>(`${this.baseUrl}/v1/dashboards/products`, { params });
  }
}
